sdks/python/apache_beam/runners/direct/direct_runner.py

```diff
@@ -327,25 +324,41 @@ def __init__(self):
   def run_pipeline(self, pipeline):
     """Execute the entire pipeline and returns an DirectPipelineResult."""
 
-    # Performing configured PTransform overrides.
-    pipeline.replace_all(_get_transform_overrides(pipeline.options))
-
     # TODO: Move imports to top. Pipeline <-> Runner dependency cause problems
     # with resolving imports when they are at top.
     # pylint: disable=wrong-import-position
+    from apache_beam.pipeline import PipelineVisitor
     from apache_beam.runners.direct.consumer_tracking_pipeline_visitor import \
       ConsumerTrackingPipelineVisitor
     from apache_beam.runners.direct.evaluation_context import EvaluationContext
     from apache_beam.runners.direct.executor import Executor
     from apache_beam.runners.direct.transform_evaluator import \
       TransformEvaluatorRegistry
+    from apache_beam.testing.test_stream import TestStream
+
+    # Performing configured PTransform overrides.
+    pipeline.replace_all(_get_transform_overrides(pipeline.options))
+
+    # If the TestStream I/O is used, use a mock test clock.
+    class _TestStreamUsageVisitor(PipelineVisitor):
+      """Visitor determining whether a Pipeline uses a TestStream."""
+
+      def __init__(self):
+        self.uses_test_stream = False
+
+      def visit_transform(self, applied_ptransform):
+        if isinstance(applied_ptransform.transform, TestStream):
+          self.uses_test_stream = True
+
+    visitor = _TestStreamUsageVisitor()
+    pipeline.visit(visitor)
```

Comment:
<!--thread_id:cc_167378324_t; commit:135fe992f11fa41ee84f83e8231995954de8aa3b; resolved:0-->
<!--section:context-quote-->
> **robertwb** wrote:
> I was just commenting that querying whether a a test stream is present requires three lines and reaching into a private variable. Instead, add
> 
> ```
> def visit_pipeline(self, pipeline):
>   self._uses_test_stream = False
>   pipeline.visit(self)
>   return self._uses_test_stream
> ```
> 
> and then you can write 
> 
> `use_test_clock = _TestStreamUsageVisitor().visit_pipeline(pipeline)`

<!--section:body-->
Sorry, it was not in response to your message, but to Charles as a general comment, that as is, the self._use_test_clock = False is still in the direct_runner.py